nrf_rpc/include/nrf_rpc.h

```diff
@@ -0,0 +1,444 @@
+ * @param handler      Callback that handles the response. In case of error
+ *                     it is undefined if the handler will be called.
+ * @param handler_data Opaque pointer that will be passed to @a handler.
+ * @returns            NRF_RPC_SUCCESS or error code from
+ *                     enum nrf_rpc_error_code.
+ */
+int nrf_rpc_cmd_send(nrf_rpc_alloc_ctx ctx, uint8_t cmd, uint8_t *packet,
+		     size_t len, nrf_rpc_handler_t handler, void *handler_data);
+
+
+/** @brief Send a command and get response directly.
+ *
+ * After successful return caller is resposible for decoding content of the
+ * response packet an call @a nrf_rpc_decoding_done just after that. After
+ * calling @a nrf_rpc_decoding_done @a rsp_packet is no longer valid.
+ *
+ * Depending on transport layer implementation this function may be slightly
+ * slower than @a nrf_rpc_cmd_send, because additional thread context switching
+ * may happen.
+ *
+ * @param remote_ep       Destination endpoint allocated @a NRF_RPC_CMD_ALLOC.
+ * @param cmd             Command id.
+ * @param packet          Packet allocated by @a NRF_RPC_CMD_ALLOC and filled
+ *                        with encoded data.
+ * @param len             Length of the packet. Can be smaller than allocated.
+ * @param[out] rsp_packet Response packet.
+ * @param[out] rsp_len    Response packet length.
+ * @returns               NRF_RPC_SUCCESS or error code from
+ *                        enum nrf_rpc_error_code.
+ */
+int nrf_rpc_cmd_rsp_send(nrf_rpc_alloc_ctx ctx, uint8_t cmd, uint8_t *packet,
+			 size_t len, const uint8_t **rsp_packet,
+			 size_t *rsp_len);
+
+
+/** @brief Send a command passing any errors to an error handler.
+ *
+ * If error occurred during sending this command it will be passed to
+ * @a nrf_rpc_error_handler instead of returning it. This form of error handling
+ * can be useful for serializing API function that has no ability to report
+ * an error in any other way, e.g. it returns void.
+ *
+ * See @a nrf_rpc_cmd_send for more details.
+ */
+void nrf_rpc_cmd_send_noerr(nrf_rpc_alloc_ctx ctx, uint8_t cmd, uint8_t *packet,
+			    size_t len, nrf_rpc_handler_t handler,
+			    void *handler_data);
+
+
+/** @brief Send an event.
+ *
+ * Sending an event always allocates a new thread from thread pool to handler
+ * it, so it should be done carefully. Seding to many events at once may
+ * consume all remote thread and as the result block other remote calls.
+ *
+ * @param remote_ep Destination endpoint allocated by @a NRF_RPC_ENT_ALLOC.
+ * @param evt       Event id.
+ * @param packet    Packet allocated by @a NRF_RPC_ENT_ALLOC and filled with
+ *                  encoded data.
+ * @param len       Length of the packet. Can be smaller than allocated.
+ * @returns         NRF_RPC_SUCCESS or error code from enum nrf_rpc_error_code.
+ */
+int nrf_rpc_evt_send(nrf_rpc_alloc_ctx ctx, uint8_t evt, uint8_t *packet,
+		     size_t len);
```

Comment:
If we decide to rename events to commands, we still have to explain the user different between those two.

Event is not simply command without response. From the caller point of view `nrf_rpc_evt_send` returns as soon as event is send to the other side. This may be even before it is received by the remote side and for sure before it is decoded on the remote side. So code on calling thread and on remote thread runs concurrently. User don't know when event completes. Acks across the IPC to release the resources are only the internal matter. Caller have no knowledge about it. Acks has not influence on the caller side with one exception: when the thread pool limit is reached then all calls are blocked until it some thread is released (via response or ack).

Command, on the other hand, waits in `nrf_rpc_cmd_send` until remote procedure completes, so the local thread stops and remote thread executes the code.

Look at the svg diagrams to compare events and commands.

I think that they need different names, because of significant differences that I described above. Even if we decide to drop `event` name the we have to think of something else to distinguish them, e.g. synchronous commands and asynchronous commands.

Failing to call `nrf_rpc_decoding_done()` in event decoder will not block the caller. It will block all the future commands and events (at lest in current rpmsg transport implementation) until event that forget to call `nrf_rpc_decoding_done()` completes.

Reply:
I already saw a need to know that ACK arrived, so I added optional `ack_handler` to the group definition parameters (not in PR yet). Something like this:
```
int my_ack_handler(uint8_t id, int return_value, void *handler_data);

NRF_RPC_GROUP_DEFINE(my_group, "my_group", my_ack_handler, &some_data);
```

I still thing that events are good way to realize asynchronous API calls, so I prefer to leave it as it is now.

Reply:
I added callback when an event on the remote end completed (based on your comment). I still want to keep events concept as the faster alternative for implementing asynchronous remote API. Do you agree to keep events?